### ISSUE TITLE
Migrate to PostCSS 8

### ISSUE THREAD

**Issue by hudochenkov** [stylelint/stylelint#4942]
Steps:

- [x] Update to latest PostCSS dependencies
- [x] Fix failing tests 
- [x] Fix failing TypeScript errors
- [x] Migrate to built-in rules to visitor API to remove deprecation warnings
- [x] Investigate impact on executing order

[Release notes](https://github.com/postcss/postcss/releases/tag/8.0.0).
[Migration guide](https://evilmartians.com/chronicles/postcss-8-plugin-migration).

---

Before we migrate we need all our dependencies be on PostCSS 8.

`dependencies` are ready:

- [x] [`@stylelint/postcss-css-in-js`](https://github.com/stylelint/postcss-css-in-js) — https://github.com/stylelint/postcss-css-in-js/pull/74 (removed)
- [x] [`@stylelint/postcss-markdown`](https://github.com/stylelint/postcss-markdown) (removed)
- [x] [`autoprefixer`](https://github.com/postcss/autoprefixer)
- [x] [`postcss`](https://github.com/postcss/postcss)
- [x] [`postcss-html`](https://github.com/stylelint/postcss-html) — https://github.com/stylelint/postcss-html/pull/1. (removed)
- [x] [`postcss-less`](https://github.com/shellscape/postcss-less)
- [x] [`postcss-media-query-parser`](https://github.com/dryoma/postcss-media-query-parser) — it doesn't use PostCSS.
- [x] [`postcss-resolve-nested-selector`](https://github.com/davidtheclark/postcss-resolve-nested-selector) — doesn't depend on PostCSS.
- [x] [`postcss-safe-parser`](https://github.com/postcss/postcss-safe-parser)
- [x] [`postcss-sass`](https://github.com/AleshaOleg/postcss-sass)
- [x] [`postcss-scss`](https://github.com/postcss/postcss-scss)
- [x] [`postcss-selector-parser`](https://github.com/postcss/postcss-selector-parser) — doesn't have PostCSS as peer or regular dependency.
- [x] [`postcss-syntax`](https://github.com/stylelint/postcss-syntax) — https://github.com/stylelint/postcss-syntax/pull/2. (removed)
- [x] [`postcss-value-parser`](https://github.com/TrySound/postcss-value-parser) — doesn't depend on PostCSS.
- [x] [`sugarss`](https://github.com/postcss/sugarss)

`devDependencies` are ready:

- [x] [`postcss-import`](https://github.com/postcss/postcss-import)

---

Update won't be easy. We need to update custom syntaxes: `@stylelint/postcss-css-in-js`, `@stylelint/postcss-markdown`. And we probably need to move into stylelint organization and update `postcss-html` and `postcss-syntax`.

---

After all that we can start updating stylelint itself.

**Comment by hudochenkov** [comment]
Out of curiosity I just bumped PostCSS to see what happen.

52 errors from TypeScript, mostly about missing imports. Probably because of new TypeScript types structure in PostCSS.

Jest hanged (it was running, but no progress was made) with this message:

```
(node:93150) UnhandledPromiseRejectionWarning: TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'Object'
    |     property 'parent' -> object with constructor 'Object'
    |     property 'nodes' -> object with constructor 'Array'
    --- index 0 closes the circle
    at stringify (<anonymous>)
    at writeChannelMessage (internal/child_process/serialization.js:117:20)
    at process.target._send (internal/child_process.js:779:17)
    at process.target.send (internal/child_process.js:677:19)
    at reportSuccess (/Users/user/projects/stylelint/node_modules/jest-worker/build/workers/processChild.js:67:11)
(node:93150) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 4399)
(node:93150) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

And this was it's status at that moment:

```
Test Suites: 5 failed, 50 passed, 55 of 339 total
```

Almost every test has this warning:

```
    console.warn
      postcss.plugin was deprecated. Migration guide:
      https://evilmartians.com/chronicles/postcss-8-plugin-migration
```

It's because every of 205 rules we have is a PostCSS plugin.

**Comment by hudochenkov** [comment]
I've migrated my [postcss-sorting](https://github.com/hudochenkov/postcss-sorting) to PostCSS 8 and it was easy (didn't use new API, though). I use following syntaxes in tests and no issues:

- `@stylelint/postcss-css-in-js`
- `postcss-html`
- `postcss-syntax`

Maybe it won't be so hard to migrate stylelint. But then we could have strange issues because of different PostCSS versions in syntaxes and stylelint itself.

**Comment by hudochenkov** [comment]
I forked [postcss-html](https://github.com/stylelint/postcss-html) and [postcss-syntax](https://github.com/stylelint/postcss-syntax) to be part of stylelint organization, so we could update them and release under `@stylelint` namespace.

**Comment by stof** [comment]
what is the status of this work ?
The update of my other postcss-based tooling in my project are blocked because I cannot really depend on 2 versions on postcss depending on whether stylelint plugins or gulp are calling postcss (or at least I don't know how that could work in a reliable way regarding dependencies)

**Comment by hudochenkov** [comment]
I believe you could continue to use stylelint as you would before. Even if you use it as a PostCSS plugin. Because PostCSS 8 support plugins made for PostCSS 7.

**Comment by niksy** [comment]
> Because PostCSS 8 support plugins made for PostCSS 7.

Is this true? It seems to me that most of the plugin updates are following documentation recommendation to use `Once` hook to make plugin compatible with PostCSS 8, and then later switch to more performant hooks.

**Comment by stof** [comment]
@niksy postcss 8 still supports the `postcss.plugin` API. But using it triggers a warning.

**Comment by hudochenkov** [comment]
I wrote down some issues we have with custom syntaxes and what to do about that in PostCSS repository https://github.com/postcss/postcss/issues/1498

**Comment by stof** [comment]
@hudochenkov is there any progress on that front ? Snyk now reports a ReDoS vulnerability in postcss, and the patch won't be backported in 7.x according to https://github.com/postcss/postcss/commit/8682b1e4e328432ba692bed52326e84439cec9e4#commitcomment-49713208

**Comment by hudochenkov** [comment]
Unfortunately, no news about that.

**Comment by ftlno** [comment]
Hmm Any progress?

**Comment by RopoMen** [comment]
This showed up today in ```npm audit``` run https://www.npmjs.com/advisories/1693

**Comment by jathak** [comment]
> Alternatively, we could pull syntaxes (and postcss-syntax) out of stylelint for [14.0.0](https://github.com/stylelint/stylelint/issues/4942). Users are welcome to stay on `13.0.0` if they need the functionality, and those that don't can upgrade to a secure version.

This would be valuable for us, both for allowing us to migrate to PostCSS 8 (we're subject to a fairly strict one-version policy that prevents us from upgrading any package that requires PostCSS 8 until all packages support it) and for simplifying our process for bringing new stylelint versions into our codebase (we currently manually patch out the syntaxes we don't care about to avoid having to import extra Node modules).

**Comment by jeddy3** [comment]
> Interestingly enough I was thinking about the same. But not all syntaxes. 

Let's pull all syntaxes so that there's a singular mechanism for using syntaxes (`--custom-syntax`). This will be simpler for both users and for us. It'll also cater to https://github.com/stylelint/stylelint/issues/4942#issuecomment-839125422, and anyone else wanting a leaner package. If we treat all syntaxes equally, we avoid having to decide which syntaxes are blessed and which aren't. We also avoid this happening again, e.g. postcss-sass is a wrapper around the gonzales-pe parser which hasn't been updated in over a year. CSS-like flavours come and go, but CSS itself will be around for the long haul. Lastly, it'll align with the behaviour of ESLint. I imagine there'll be a move to more narrowly scoped syntaxes, e.g. [`vue-eslint-parser`](https://www.npmjs.com/package/vue-eslint-parser). It'll be easier to write a PostCSS syntax for just Vue's needs than try to wrangle a syntax like postcss-css-in-js that tries to work for every CSS-in-JS flavour under the sun, many of which have come and gone already.

I'll create a new issue for removing the syntaxes so that we can make this issue about migrating the rules.

---

How do we make our rules compatible with PostCSS 8? Am I right in thinking that [step 2 of the migration guide]( https://evilmartians.com/chronicles/postcss-8-plugin-migration#step-2-use-the-updated-api) will be enough for now? We can then more fully embrace the new visitor API later down the line.

If so, it looks like [this codemod](https://github.com/coderaiser/putout/tree/master/packages/plugin-postcss) will help.

**Comment by ludofischer** [comment]
> How do we make our rules compatible with PostCSS 8? Am I right in thinking that [step 2 of the migration guide](https://evilmartians.com/chronicles/postcss-8-plugin-migration#step-2-use-the-updated-api) will be enough for now? We can then more fully embrace the new visitor API later down the line.
> 
> If so, it looks like [this codemod](https://github.com/coderaiser/putout/tree/master/packages/plugin-postcss) will help.

From my experience on cssnano, using Once ensures all warnings about deprecated APIs disappear. 
Moving to the visitor API causes issues if  the code assumes (even implicitly) the plugins run in a definite order. Issues go as far as infinite loops if plugins 'undo' each other's work. If stylelint does not modify the AST at all, the order between eslint plugins might not matter at all since a plugin cannot alter what the other plugins `see`. But you need to take into account that users might add other plugins which could interfere with the result.
The advantage of the visitor API is performance and possibly terser code (since you don't need to check which type of node you are visiting)

**Comment by jeddy3** [comment]
Thanks for the heads up!

> If stylelint does not modify the AST at all

It does when autofixing. Will using `Once` be OK for us?

> The advantage of the visitor API is performance

Yep, later down the line, it'd be great to [benchmark](https://stylelint.io/developer-guide/rules#improve-the-performance-of-a-rule) the performance gains we can bring to some of our heavier rules.

**Comment by hudochenkov** [comment]
> Moving to the visitor API causes issues if the code assumes (even implicitly) the plugins run in a definite order. 

We rely on order of rules for autofixing. E. g. indentation runs the last because other rules could change something indentation wise and we wanted to fix that.

**Comment by jeddy3** [comment]
Will `OnceExit` on `indentation suffice as a workaround in PostCSS 8? It's the one rule we've stuck at the bottom of the [list of rules](https://github.com/stylelint/stylelint/blob/master/lib/rules/index.js#L398).

**Comment by hudochenkov** [comment]
One more thing about order. I believe we run plugins before core rules. This way when `stylelint-order` changes order of things, core rules fix `*-empty-line-before` issues that could arise.

I think we need to implement mechanics similar to ESLint — run fix until all fixable problems are fixed, but no more than 10 times.

**Comment by ludofischer** [comment]
> Will `OnceExit` on `indentation` suffice as a workaround in PostCSS 8? It's the one rule we've stuck at the bottom of the [list of rules](https://github.com/stylelint/stylelint/blob/master/lib/rules/index.js#L398).

In my experience, it's better to stick with either `Once` or `OnceExit`. If you define every plugin with `Once` or every plugin with `OnceExit` they run in the order you pass them to `postcss`. If you run mix `Once` and `OnceExit` you have less control, as a third-party `Once` plugin could jump in between your `Once` and `OnceExit` rules.

**Comment by jeddy3** [comment]
We've got the ball rolling on the `14.0.0` release (https://github.com/stylelint/stylelint/issues/5205#issue-837034702), and this issue is now unblocked.

I've added a checklist to [the top post](https://github.com/stylelint/stylelint/issues/4942#issue-702121905) for the update to PostCSS 8. There's a [Draft PR](https://github.com/stylelint/stylelint/pull/5304), and [associated branch](https://github.com/stylelint/stylelint/tree/update-to-postcss-8), for this work. 

There are [6 failing tests](https://github.com/stylelint/stylelint/pull/5304/checks?check_run_id=2589619871). Most seem to be around a change in parser behaviour for custom property sets. I suggest we pick these up in 3 separate pull requests:

- [ ] FAIL` lib/__tests__/disableRanges.test.js`
- [ ] FAIL `lib/rules/string-no-newline/__tests__/index.js`
- [ ] The rest (which are connected to custom properties)

Once we've patched up the failing tests and TypeScript errors, we can migrate the built-in rules to the visitor API to remove the deprecation warnings. Lastly, we'll need to investigate the impact of this change on the execution order.

Please shout out if you'd like to pick up one of the failing tests pull requests, or start looking at the TypeScript errors. Remember to branch off `update-to-postcss-8`.

**Comment by hudochenkov** [comment]
Before migrating a rule to visitor API we would need to change how we run rules. Currently we going around PostCSS and [call rule function directly](https://github.com/stylelint/stylelint/blob/b01ed25dfa3e8231a976eef76bd5e81fb535b1e9/lib/lintPostcssResult.js#L103-L110). It will not work anymore, because now PostCSS plugin is a different function.

We need to use PostCSS API to call plugins now. E. g.:

```js
await postcss([plugin(pluginOptions)]).process(root, {
    from: result.from,
});
```

**Comment by jeddy3** [comment]
The last failing test we need to fix is:

```console
 FAIL  lib/__tests__/disableRanges.test.js
  ● Less // disable next-line comment (with multi-line description)

    expect(received).toEqual(expected) // deep equality

    - Expected  - 3
    + Received  + 3

      Object {
    -   "description": "Long-winded description",
    -   "end": 5,
    -   "start": 5,
    +   "description": undefined,
    +   "end": 39,
    +   "start": 39,
        "strictEnd": true,
        "strictStart": true,
      }

      1000 | 
      1001 |    delete actualMutable.comment;
    > 1002 |    expect(actualMutable).toEqual(expected);
           |                          ^
      1003 | }
      1004 | 
      1005 | function testDisableRanges(source, options) {

      at expectDisableRange (lib/__tests__/disableRanges.test.js:1002:24)
      at expectDisableRanges (lib/__tests__/disableRanges.test.js:993:4)
      at Object.<anonymous> (lib/__tests__/disableRanges.test.js:968:2)
```

For this test:

```js
it('Less // disable next-line comment (with multi-line description)', async () => {
	const result = await testDisableRangesLess(
		`a {
			// stylelint-disable-next-line declaration-no-important
			// --
			// Long-winded description
			color: pink !important;
		}`,
	);

	expectDisableRanges(result, {
		all: [],
		'declaration-no-important': [
			{
				start: 5,
				end: 5,
				strictStart: true,
				strictEnd: true,
				description: 'Long-winded description',
			},
		],
	});
});
```

It's the workaround we added to merge adjacent double slash comments into one.

`comment.source.end.line` jumps to `69` at some point, which doesn't seem right. It may be an issue upstream in the less parser or one we need to address in stylelint.

@jathak do you have any ideas? I think you worked on this workaround at the time.

**Comment by jeddy3** [comment]
I think we can, if anyone fancies it, pick up the TypeScript errors while we look into this last failing test?

**Comment by kaaax0815** [comment]
This would fix: [CVE-2021-23368](https://github.com/advisories/GHSA-hwj9-h5mp-3pm3)
```
stylelint@13.13.1 requires postcss@^7.0.32 via a transitive dependency on autoprefixer@9.8.6
stylelint@13.13.1 requires postcss@^7.0.35
stylelint@13.13.1 requires postcss@^7.0.14 via postcss-less@3.1.4
stylelint@13.13.1 requires postcss@^7.0.26 via postcss-safe-parser@4.0.2
stylelint@13.13.1 requires postcss@^7.0.21 via a transitive dependency on postcss-sass@0.4.4
stylelint@13.13.1 requires postcss@^7.0.6 via postcss-scss@2.1.1
stylelint@13.13.1 requires postcss@^7.0.2 via sugarss@2.0.0
```

**Comment by jeddy3** [comment]
That's correct. The [Draft PR](https://github.com/stylelint/stylelint/pull/5304) has no PostCSS@7 within its tree.

**Comment by jeddy3** [comment]
> Before migrating a rule to visitor API we would need to change how we run rules. Currently we going around PostCSS and call rule function directly.

As our rules and plugins are rule functions, is there a way we can adapt them `Once: {}` PostCSS 8 plugins _on the fly_?

I'm not very familiar with the engine as I've been more focused on the rules since the beginning of the project. So, if anyone fancies picking up this migration to PostCSS 8, please go ahead as I don't think I'll have time myself.

**Comment by hudochenkov** [comment]
> As our rules and plugins are rule functions, is there a way we can adapt them `Once: {}` PostCSS 8 plugins _on the fly_?

If it's possible and we do this we need to find a way how to support rules and plugins, which use visitor API.

**Comment by jeddy3** [comment]
Yep. I figured we could get the existing rules and plugins working first and then look into the visitor API. I think it's OK if we release without support for the visitor API. We can update our [rule](https://stylelint.io/developer-guide/rules) and [plugin](https://stylelint.io/developer-guide/plugins) docs to say that the stylelint rule function will be placed within the PostCSS `Once` context, and therefore a stylelint rule/plugin can't use the other visitors from the API.

Better to get something secure out-the-door than try to support of a type of stylelint rule or plugin that's yet to be written. Of course, if there's a way of doing both then that would be great. But I think we should prioritise the former over the latter.

**Comment by hudochenkov** [comment]
I'll be fixing type issues and then try to continue migration to PostCSS 8. I've already started doing some work in a new branch, which is branched out from `update-to-postcss-8`.

**Comment by jeddy3** [comment]
Closed by https://github.com/stylelint/stylelint/pull/5304

**Comment by jeddy3** [comment]
For anyone following this issue, there's still [lots to do](https://github.com/stylelint/stylelint/issues/5205#issue-837034702) before we can release `14.0.0`. Please consider [contributing](https://stylelint.io/contributing) to any of the [outstanding issues](https://github.com/stylelint/stylelint/issues/5205#issue-837034702), if you'd like to help make the release happen.

**Comment by Stanzilla** [comment]
@jeddy3 any plans for an interim 13.x release just to fix the vulnerabilities and deprecation warnings?

**Comment by jeddy3** [comment]
A 13.x release isn't possible because we had to make breaking changes, including [removing the bundled syntaxes](https://github.com/stylelint/stylelint/issues/5289), e.g. CSS-in-JS etc, to make the migration PostCSS@8 possible.

There is also a vulnerability in another dependency that is now ESM-only, so we need to migrate to ESM to fix all vulnerabilities.

The best thing everyone can do is help us work through the remaining issues so that we release `14.0.0` as soon as possible.

**Comment by hudochenkov** [comment]
PostCSS v7 was released with back-port of ReDoS fix. So stylelint users on current version of stylelint should see warning go away after running `npm audit fix` (it could take few days until `npm audit` will know about the fix).

https://twitter.com/PostCSS/status/1403351574110511106
